pkg/s3/fetch.go

```diff
@@ -0,0 +1,196 @@
+package s3
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+type Fetcher struct {
+	Logger *log.Logger
+}
+
+func (f *Fetcher) Fetch(dir string, s3URL string, s3AccessKey string, s3SecretKey string, s3Bucket string, s3File string, s3ForcePathStyle string, s3Region string) error {
+	blob, err := url.Parse(s3URL)
+	if err != nil {
+		return err
+	}
+
+	option := session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}
+
+	forcePath, err := strconv.ParseBool(s3ForcePathStyle)
+	if err != nil {
+		forcePath = false
+	}
+
+	region := s3Region
+	if region == "" {
+		region = "us-west-2"
+	}
+
+	option.Config = aws.Config{
+		Endpoint:         aws.String(s3URL),
+		S3ForcePathStyle: aws.Bool(forcePath),
+		Credentials:      credentials.NewStaticCredentials(s3AccessKey, s3SecretKey, ""),
+		Region:           aws.String(region),
+	}
+
+	sess, err := session.NewSessionWithOptions(option)
+	downloader := s3manager.NewDownloader(sess)
+
+	file, err := ioutil.TempFile("", "")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(file.Name())
+
+	_, err = downloader.Download(file, &s3.GetObjectInput{
+		Bucket: aws.String(s3Bucket),
+		Key:    aws.String(s3File),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	fileType, err := classifyFile(file.Name())
+	if err != nil {
+		return err
+	}
+
+	switch fileType {
+	case "application/zip":
+		err = extractZip(file, dir)
+	case "application/x-gzip":
+		r, err := os.Open(file.Name())
+		if err != nil {
+			return err
+		}
+
+		gzf, err := gzip.NewReader(r)
+		if err != nil {
+			return err
+		}
+
+		err = extractTar(gzf, dir)
+	case "application/octet-stream":
+		r, err := os.Open(file.Name())
+		if err != nil {
+			return err
+		}
+		err = extractTar(r, dir)
+	}
+
+	if err != nil {
+		return err
+	}
+	f.Logger.Printf("Successfully downloaded %s/%s/%s in path %q", blob.Host+blob.Path, s3Bucket, s3File, dir)
+	return nil
+}
+
+func extractTar(reader io.Reader, dir string) error {
```

Comment:
maybe time to extract these 3 methods: `extractTar`, `extractZip`, `classifyFile`